### ISSUE TITLE
getHeartRateSamples Freezes UI / Slow Performance

### ISSUE THREAD

**Issue by shmkane** [agencyenterprise/react-native-health#157]
**Description**

**Summary**: The react-native-health's `getHeartRateSamples` function freezes the UI for significant time while fetching data, particularly heart rate.

An [article](https://www.zdnet.com/article/apple-watch-heart-rate-every-10-minutes-motion/) claims that the Apple Watch Series 5 samples heart rate every 10 minutes- more if an activity event is in progress. From my own data/testing, I see approximately 10-20 samples every hour. For example, Img 1 below, I count 11. But this is on the low end.

<img src="https://user-images.githubusercontent.com/2586027/133135034-0edb68c4-6a14-47f5-9bc4-280f903fd3cc.png" height="300">
Img 1.  <br />
  <br />
  <br />

According to my own data, requesting data back to April 4th of this year took around *11 seconds*! During which time the UI remains almost completely unresponsive!

I went through and tested other dates and my findings were close to linear (Img 2 below), reported in Table 1 below.

|Start Date | Number days | Data Points | Time to un-freeze (ms)|
|-----------|--------------|------------|------------------------|
|'2021-05-06'| 112| 38728| 11865 (11.8s)|
|'2021-02-01'| 206| 59864|  18559 (18.5s)|
|'2020-02-01'| 572| 110177| 49504 (49.5s)|

Table 1. <br />

The end date was left as default to "today", which at the time of testing this, was August 26th 2021.

**To Reproduce**

I used the sample code provided in the README.md of this repository. To be verbose, I've pasted it here
```

AppleHealthKit.initHealthKit(permissions, (error: string) => {
  const options = {
   // ----------DATE SET HERE 
    startDate: new Date(2020, 1, 1).toISOString(), 
  }

 // ---------"TIMER" WAS STARTED HERE
  AppleHealthKit.getHeartRateSamples(
    options,
    (callbackError: string, results: HealthValue[]) => {
     //--------------"TIMER" WAS ENDED HERE
    },
  )
})

```

This was called at the top level component in my App.tsx.

**Expected behavior**
I expect the UI to not freeze up during the time the data is being fetched from AHK.

**Screenshots**

![image](https://user-images.githubusercontent.com/2586027/133136762-97c6d198-ba40-4e2b-bce1-39ece60ffb5a.png)
Img 2.

**Smartphone (please complete the following information):**

- Device: iPhone 11 Pro
- OS: 15.0 (19A5340a)

**Additional context**
Would love to know if there's a better approach to fetching large dates of data. My current approach is to split the date range into 1 day intervals and call `getHeartRateSamples` for each 1 day interval date range. I'd hate for this to be the only solution.

TL;DR - AppleHealthKit when there’s a LOT of data. ~100 days of data = ~38,000 HR data points = ~11 seconds UI freeze time

Edit: Using react-native-health: 1.10.0

**Comment by macelai** [comment]
Hey,

We had some problems with other queries as well, maybe if you see them you could have some insights #113

**Comment by shmkane** [comment]
> Hey,
> 
> We had some problems with other queries as well, maybe if you see them you could have some insights #113

Thanks for the reply.
I see that the PR you linked had a change that decreased response time by specifying a period parameter in the `NSPredicate`, but I don't see how this helps the issue I'm having. I'm currently using version 1.10.0, which I assume has this fix implemented, so I'd expect a quicker response time as well.

Furthermore, in the requests I'm making, I am also specifying the startDate and endDate.
```
const options = {
   // ----------DATE SET HERE 
    startDate: new Date(2020, 1, 1).toISOString(), 
  }
```


Here's a screen cap of what I see if I try to make several smaller (1 day) "batches".  Attaching a code snippet that produces this output below


https://user-images.githubusercontent.com/2586027/133152023-922a19c5-aea7-4c6c-a0df-deca7cb9ad5b.mov


```
  const END_DATE = new Date();
  const START_DATE = new Date(2021, 7, 1);

  // returns an array of dates 1 day apart, starting at arg1, ending at arg2
  const batches = dateRangeBatcher(START_DATE, END_DATE);
  const numBatches = batches.length;

  for (let i = 0; i < numBatches; i++) {
    const batch: BatchType = batches[i];

    const options = {
      startDate: batch.startBatch.toISOString(),
      endDate: batch.endBatch.toISOString(),
    };

    // Start Timer
    const startTime = new Date().getTime();
    AppleHealthKit.getHeartRateSamples(
      options,
      async (_callbackError: string, _results: HealthValue[]) => {
        // End Timer
        const endTime = new Date().getTime();

        console.log(
          batch.startBatch.toDateString(),
          " to ",
          batch.endBatch.toDateString(),
          "took:"
        );

        console.log("Took", (endTime - startTime) / 1000, "seconds");
      }
    );
  }
  ```

**Comment by kristerkari** [comment]
@shmkane @macelai 

The probable reason for the UI freezing is that the queries are performed on the main thread ([RCTAppleHealthKit+Queries.m](https://github.com/agencyenterprise/react-native-health/blob/master/RCTAppleHealthKit/RCTAppleHealthKit%2BQueries.m)), which then blocks things in the UI since it's running on the same thread.

There is a fix in #176, which moves the queries off the main thread to another thread (another queue). I have tested those changes and the UI freezing was completely fixed for me.
